From 6118b2bd885946bb86ebec33282cbff75a371769 Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Thu, 20 Jul 2023 19:21:40 +0800
Subject: [PATCH 06/53] [Huawei] mm/sharepool: Delete useless wrap functions in
 sp_alloc

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

Delete sp_alloc_mmap/unmap/populate() and unwrap those code at where
they are called.

This patch is to facilitate future reconstruction. No process change
is made.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 96 +++++++++++++++++++------------------------------
 1 file changed, 36 insertions(+), 60 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index 61526892b2af..484006ab7240 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -2228,16 +2228,36 @@ static int sp_alloc_prepare(unsigned long size, unsigned long sp_flags,
 	return 0;
 }
 
-static void sp_alloc_unmap(struct mm_struct *mm, struct sp_area *spa,
-	struct sp_group_node *spg_node)
+static void sp_alloc_fallback(struct sp_area *spa, struct sp_alloc_context *ac)
 {
-	__sp_free(spa->spg, spa->va_start, spa->real_size, mm);
+	if (ac->file == ac->spg->file) {
+		ac->state = ALLOC_NOMEM;
+		return;
+	}
+
+	if (!(ac->sp_flags & SP_HUGEPAGE_ONLY)) {
+		ac->file = ac->spg->file;
+		ac->size_aligned = ALIGN(ac->size, PAGE_SIZE);
+		ac->sp_flags &= ~SP_HUGEPAGE;
+		ac->state = ALLOC_RETRY;
+		__sp_area_drop(spa);
+		return;
+	}
+	ac->state = ALLOC_NOMEM;
 }
 
-static int sp_alloc_mmap(struct mm_struct *mm, struct sp_area *spa,
-	struct sp_group_node *spg_node, struct sp_alloc_context *ac)
+static long sp_mbind(struct mm_struct *mm, unsigned long start, unsigned long len,
+		nodemask_t *nodemask)
 {
-	int ret = 0;
+	return __do_mbind(start, len, MPOL_BIND, MPOL_F_STATIC_NODES,
+			nodemask, MPOL_MF_STRICT, mm);
+}
+
+static int __sp_alloc_mmap_populate(struct mm_struct *mm, struct sp_area *spa,
+	struct sp_group_node *spg_node, struct sp_alloc_context *ac,
+	nodemask_t *nodemask)
+{
+	int ret;
 	unsigned long mmap_addr;
 	/* pass through default permission */
 	unsigned long prot = PROT_READ | PROT_WRITE;
@@ -2262,9 +2282,9 @@ static int sp_alloc_mmap(struct mm_struct *mm, struct sp_area *spa,
 	mmap_addr = sp_mmap(mm, spa_file(spa), spa, &populate, prot, &vma);
 	if (IS_ERR_VALUE(mmap_addr)) {
 		up_write(&mm->mmap_lock);
-		sp_alloc_unmap(mm, spa, spg_node);
+		__sp_free(spa->spg, spa->va_start, spa_size(spa), mm);
 		pr_err("sp mmap in allocation failed %ld\n", mmap_addr);
-		return PTR_ERR((void *)mmap_addr);
+		return (int)mmap_addr;
 	}
 
 	ac->populate = populate;
@@ -2276,56 +2296,6 @@ static int sp_alloc_mmap(struct mm_struct *mm, struct sp_area *spa,
 		vma->vm_page_prot = __pgprot(((~PTE_RDONLY) & vma->vm_page_prot.pgprot) | PTE_DIRTY);
 	up_write(&mm->mmap_lock);
 
-	return ret;
-}
-
-static void sp_alloc_fallback(struct sp_area *spa, struct sp_alloc_context *ac)
-{
-	if (ac->file == ac->spg->file) {
-		ac->state = ALLOC_NOMEM;
-		return;
-	}
-
-	if (!(ac->sp_flags & SP_HUGEPAGE_ONLY)) {
-		ac->file = ac->spg->file;
-		ac->size_aligned = ALIGN(ac->size, PAGE_SIZE);
-		ac->sp_flags &= ~SP_HUGEPAGE;
-		ac->state = ALLOC_RETRY;
-		__sp_area_drop(spa);
-		return;
-	}
-	ac->state = ALLOC_NOMEM;
-}
-
-static int sp_alloc_populate(struct mm_struct *mm, struct sp_area *spa,
-			     struct sp_alloc_context *ac)
-{
-	/*
-	 * We are not ignoring errors, so if we fail to allocate
-	 * physical memory we just return failure, so we won't encounter
-	 * page fault later on, and more importantly sp_make_share_u2k()
-	 * depends on this feature (and MAP_LOCKED) to work correctly.
-	 */
-	return do_mm_populate(mm, spa->va_start, ac->populate, 0);
-}
-
-static long sp_mbind(struct mm_struct *mm, unsigned long start, unsigned long len,
-		nodemask_t *nodemask)
-{
-	return __do_mbind(start, len, MPOL_BIND, MPOL_F_STATIC_NODES,
-			nodemask, MPOL_MF_STRICT, mm);
-}
-
-static int __sp_alloc_mmap_populate(struct mm_struct *mm, struct sp_area *spa,
-	struct sp_group_node *spg_node, struct sp_alloc_context *ac,
-	nodemask_t *nodemask)
-{
-	int ret;
-
-	ret = sp_alloc_mmap(mm, spa, spg_node, ac);
-	if (ret < 0)
-		return ret;
-
 	if (!ac->have_mbind) {
 		ret = sp_mbind(mm, spa->va_start, spa->real_size, nodemask);
 		if (ret < 0) {
@@ -2336,7 +2306,13 @@ static int __sp_alloc_mmap_populate(struct mm_struct *mm, struct sp_area *spa,
 		ac->have_mbind = true;
 	}
 
-	ret = sp_alloc_populate(mm, spa, ac);
+	/*
+	 * We are not ignoring errors, so if we fail to allocate
+	 * physical memory we just return failure, so we won't encounter
+	 * page fault later on, and more importantly sp_make_share_u2k()
+	 * depends on this feature (and MAP_LOCKED) to work correctly.
+	 */
+	ret = do_mm_populate(mm, spa->va_start, ac->populate, 0);
 	if (ret) {
 		if (unlikely(fatal_signal_pending(current)))
 			pr_warn_ratelimited("allocation failed, current thread is killed\n");
@@ -2391,7 +2367,7 @@ static int sp_alloc_mmap_populate(struct sp_area *spa,
 	/* use the next mm in proc list as end mark */
 	if (!list_is_last(&spg_node->proc_node, &spa->spg->proc_head))
 		end_mm = list_next_entry(spg_node, proc_node)->master->mm;
-	sp_alloc_unmap(end_mm, spa, spg_node);
+	__sp_free(spa->spg, spa->va_start, spa_size(spa), end_mm);
 
 	/*
 	 * Sometimes do_mm_populate() allocates some memory and then failed to
-- 
2.17.1

